### PR TITLE
Fix null reference exception for Streaming null object

### LINKED ISSUE #13191: ChannelReader<T> method in Hub with CancellationToken throws
```
public ChannelReader<ApiTypedData<LogInResponseType>> LogIn(string username, byte[] passwordHash, string baseUrl, string fastLoginPassword, CancellationToken cancellationToken)
{
    var channel = Channel.CreateUnbounded<ApiTypedData<LogInResponseType>>();
    _ = LogInAsync(channel.Writer, username, passwordHash, baseUrl, fastLoginPassword, cancellationToken);
    return channel.Reader;
}

async Task LogInAsync(ChannelWriter<ApiTypedData<LogInResponseType>> writer, string username, byte[] passwordHash, string baseUrl, string fastLoginPassword, CancellationToken cancellationToken)
{
    // ...
}
```
Declaring the method above causes the following exception to be thrown when invoking in the browser (JavaScript):

```
fail: Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher[8]
      Failed to invoke hub method 'LogIn'.
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Object.GetType()
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.Invoke(HubMethodDescriptor descriptor, HubConnectionContext connection, HubMethodInvocationMessage hubMethodInvocationMessage, Boolean isStreamedInvocation)
Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher:Error: Failed to invoke hub method 'LogIn'.
```

JavaScript:

```
this.hubConnection.stream("LogIn", username, passwordHash, location.origin, null).subscribe({
    next: (item) => {
        console.log("next", item);
    },
    complete: () => {
        console.log("complete")
    },
    error: (err) => {
        console.log("err", err);
    }});
```

**If I remove the CancellationToken parameter from the back-end code, it works, and the exception is not thrown.**

I have configured my hub to use MessagePack.

`<PackageReference Include="Microsoft.AspNetCore.SignalR.Protocols.MessagePack" Version="1.1.5" />`

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes #13191

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Thanks for the PR @Kahbazi!

Unfortunately I think the fix is a little more complicated than this.

Now if someone passes `null` as an argument they'll fall through to the if statement that says `// This should never happen`, we instead would want the argument to be set on the arguments array.

Also, we should add a test to verify that the scenario is fixed and that we don't regress it in the future! The test would likely be similar to https://github.com/aspnet/AspNetCore/blob/7b0b2980dc6ffb860f2bcd71868d98d85c7d35b6/src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs#L3524

**Comment by Kahbazi** [comment]
@BrennanConroy If it's possible I'd like to give it another try to fix this.
I have couple of question 

- This was the original javascript which causes the exception 

`this.hubConnection.stream("LogIn", username, passwordHash, location.origin, null).subscribe(...)`

and the workaround was to put an empty string instead of null

`this.hubConnection.stream("LogIn", username, passwordHash, location.origin, "").subscribe(...)`

Shouldn't the correct way to use javascript client to just omit the last parameter?

`this.hubConnection.stream("LogIn", username, passwordHash, location.origin).subscribe(...)`

- If I'm not wrong the current code has a bug whether the null is sent for cancellation token or any parameter so What should happen when there's a null argument for types like int, bool, DateTime? Should I put the default value on the arguments array or throw an exception?

- What should happen if the number of arguments coming from client doesn't match the number on the server?

**Comment by BrennanConroy** [comment]
>If it's possible I'd like to give it another try to fix this.

Of course! I was just giving suggestions so that we can approve and merge this.

>Shouldn't the correct way to use javascript client to just omit the last parameter?

No, in this case their method was `public ChannelReader<ApiTypedData<LogInResponseType>> LogIn(string username, byte[] passwordHash, string baseUrl, string fastLoginPassword, CancellationToken cancellationToken)` so they want to pass 4 arguments (userName, passwordHash, location.origin, and null).

>What should happen when there's a null argument for types like int, bool, DateTime? Should I put the default value on the arguments array or throw an exception?

That should be handled by the underlying protocol.

>What should happen if the number of arguments coming from client doesn't match the number on the server?

Also handled by the underlying protocol.

**Comment by Kahbazi** [comment]
> No, in this case their method was public ChannelReader<ApiTypedData<LogInResponseType>> LogIn(string username, byte[] passwordHash, string baseUrl, string fastLoginPassword, CancellationToken cancellationToken) so they want to pass 4 arguments (userName, passwordHash, location.origin, and null).

Oh my bad, I was miscounting and thought the last parameter (null) was for cancellation token. Everything makes much more sense now.

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs [hunk 1]
```diff
@@ -3551,6 +3551,33 @@ public async Task StreamHubMethodCanAcceptCancellationTokenAsArgumentAndBeTrigge
...
+                {
+                    builder.AddSingleton(tcsService);
+                }, LoggerFactory);
+                var connectionHandler = serviceProvider.GetService<HubConnectionHandler<LongRunningHub>>();
+
+                using (var client = new TestClient())
+                {
+                    var connectionHandlerTask = await client.ConnectAsync(connectionHandler).OrTimeout();
+
+                    var streamInvocationId = await client.SendStreamInvocationAsync(nameof(LongRunningHub.StreamNullableParameter),5, null).OrTimeout();
```
Comment (BrennanConroy) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]:
```suggestion
                    var streamInvocationId = await client.SendStreamInvocationAsync(nameof(LongRunningHub.StreamNullableParameter), 5, null).OrTimeout();
```

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs [hunk 2]
```diff
@@ -3551,6 +3551,33 @@ public async Task StreamHubMethodCanAcceptCancellationTokenAsArgumentAndBeTrigge
...
+                    // Wait for the stream method to start
+                    var firstArgument = await tcsService.StartedMethod.Task.OrTimeout();
+                    Assert.Equal(5, firstArgument);
+
+                    var secondArgument =await tcsService.EndMethod.Task.OrTimeout();
```
Comment (BrennanConroy) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]:
```suggestion
                    var secondArgument = await tcsService.EndMethod.Task.OrTimeout();
```

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 3]
```diff
@@ -265,7 +265,8 @@ private Task ProcessStreamItem(HubConnectionContext connection, StreamItemMessag
                         for (var parameterPointer = 0; parameterPointer < arguments.Length; parameterPointer++)
                         {
                             if (hubMethodInvocationMessage.Arguments.Length > hubInvocationArgumentPointer &&
-                                descriptor.OriginalParameterTypes[parameterPointer].IsAssignableFrom(hubMethodInvocationMessage.Arguments[hubInvocationArgumentPointer].GetType()))
+                                (hubMethodInvocationMessage.Arguments[hubInvocationArgumentPointer] == null ||
```
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: Need to actually test the fix by having a hub method that accepts a `CancellationToken`!

Just add another test similar to the one you've already added.

Comment (BrennanConroy) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: FYI, I like to remove the fix after making a test and make sure the test actually fails when the fix is removed.

Comment (Kahbazi) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: Of course! I actually did write the test with `CancellationToken` first, but I remove it at last minute. I forgot that it's the `CancellationToken` that makes `HasSyntheticArguments` true.

File: src/SignalR/server/SignalR/test/Internal/TypedClientBuilderTests.cs [hunk 4]
```diff
@@ -92,7 +92,7 @@ public async Task SupportsCancellationToken()
                 send1 =>
                 {
                     Assert.Equal("Method", send1.Method);
-                    Assert.Equal(1, send1.Arguments.Length);
+                    Assert.Single(send1.Arguments);
```
Comment (BrennanConroy) [inline src/SignalR/server/SignalR/test/Internal/TypedClientBuilderTests.cs]: Haha, thanks, I've been meaning to fix those warnings 😆

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs [hunk 5]
```diff
@@ -3551,6 +3551,64 @@ public async Task StreamHubMethodCanAcceptCancellationTokenAsArgumentAndBeTrigge
...
+                    var firstArgument = await tcsService.StartedMethod.Task.OrTimeout();
+                    Assert.Equal(5, firstArgument);
+
+                    var secondArgument = await tcsService.EndMethod.Task.OrTimeout();
+                    Assert.Null(secondArgument);
+                }
+            }
+        }
+
+
```
Comment (BrennanConroy) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: nit: remove this extra line

**Review by BrennanConroy — APPROVED** [review]
Thanks for the PR!

I want to give this a quick local test before merging and give @halter73 a chance to take a look if he'd like.

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs (hunk 3 above)
Comment (halter73) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: What if the parameter isn't a nullable type?

Comment (Kahbazi) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: If I'm not mistaken, json serializer would throw an exception when the parameter is non-nullable and the value is null, and it never reaches this code.

File: src/SignalR/server/SignalR/test/Internal/TypedClientBuilderTests.cs (hunk 4 above)
Comment (Kahbazi) [reply src/SignalR/server/SignalR/test/Internal/TypedClientBuilderTests.cs]: Well I was the one who commit that code in the first place

**Review by halter73 — APPROVED** [review]

**Pull request merged**